File: src/Asset/Vns.Erp.Core.Asset.VbAssetGUI/frmThongTinTaiSan_DanhSach.vb
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraGrid.Views.Grid

Imports Vns.Erp.Core.Asset.Domain
Imports Vns.Erp.Core.Asset.Service.Interface

Public Class frmThongTinTaiSan_DanhSach

#Region "Property"
    Private _DmLoaitaisanService As IDmLoaitaisanService
    Public Property DmLoaitaisanService As IDmLoaitaisanService
        Get
            Return _DmLoaitaisanService
        End Get
        Set(ByVal value As IDmLoaitaisanService)
            _DmLoaitaisanService = value
        End Set
    End Property
    Private _DmTscdService As IDmTscdService
    Public Property DmTscdService As IDmTscdService
        Get
            Return _DmTscdService
        End Get
        Set(ByVal value As IDmTscdService)
            _DmTscdService = value
        End Set
    End Property
#End Region

#Region "Variables"

    Dim DM_TSCD_ID As Guid
    Dim TSCD_new As DmTscd
#End Region

#Region "Private functions"

    Public Sub New()
        InitializeComponent()
        TNCommon.setIconControl(Me)
    End Sub

    Private Sub BindData()
        Dim DON_VI_ID As Guid = Generals.DON_VI.Id
        Dim objColumnDefine As ArrayList

        'Dim dt_DM_LoaiTS As DataTable = DM_LoaitaisanBO.Instance.GetAll.Tables(0)
        Dim dt_DM_LoaiTS As List(Of DmLoaitaisan) = New List(Of DmLoaitaisan)

        dt_DM_LoaiTS.AddRange(_DmLoaitaisanService.GetAllSortByKyHieu())

        Dim dr As New DmLoaitaisan
        dr.KyHieu = ""
        dr.Id = Null.NullGuid
        dr.TenLoaitaisan = "<Không xác định>"
        dt_DM_LoaiTS.Insert(0, dr)

        objColumnDefine = New ArrayList
        objColumnDefine.Add(New LookupEdit_ColumnInfo("KyHieu", "Ký hiệu"))
        objColumnDefine.Add(New LookupEdit_ColumnInfo("TenLoaitaisan", "Tên loại tài sản"))
        GridLookupEdit_BindData(grlLoaiTSCD, dt_DM_LoaiTS, "KyHieu", "Id", COMBO_ITEM_BLANK, "TenLoaitaisan", txttenLoaiTS, objColumnDefine)
    End Sub

    'bind data vao grid
    Private Sub Load_GRV_TSCD()
        Dim lst_TSCD As IList(Of DmTscd) = New List(Of DmTscd)
        lst_TSCD = _DmTscdService.GetbyParameter((dteTuNgay.DateTime), (dteDenNgay.DateTime), txtTenTS.Text.Trim, VnsConvert.CGuid(grlLoaiTSCD.EditValue), Generals.DON_VI.Id)
        grcTSCD.DataSource = lst_TSCD
        'grvTSCD.BestFitColumns()
    End Sub

    Private Sub Create()
        Dim frm As frmThongTinTSCD_ChiTiet = CType(ObjectFactory.GetObject("frmThongTinTSCD_ChiTiet"), frmThongTinTSCD_ChiTiet)
        TSCD_new = frm.Show_Form(Nothing, DataInputState.AddMode)
        If TSCD_new IsNot Nothing Then
            Me.Load_GRV_TSCD()
            grv_focus()
        End If
    End Sub

    Private Sub Edit()
        If grvTSCD.RowCount <> 0 Then
            Dim i As Integer = grvTSCD.FocusedRowHandle
            'DM_TSCD_ID = Grid_GetDataRowItemString(grvTSCD, "Id")
            DM_TSCD_ID = CType(grvTSCD.GetRow(i), DmTscd).Id

            If (Grid_GetDataRowItemInteger(grvTSCD, "GIAM_TSCD") = 1) Then
                Message_Information("Tài sản đã khai báo giảm")
                Exit Sub
            End If

            Dim frm As frmThongTinTSCD_ChiTiet = CType(ObjectFactory.GetObject("frmThongTinTSCD_ChiTiet"), frmThongTinTSCD_ChiTiet)
            TSCD_new = frm.Show_Form(DM_TSCD_ID, DataInputState.EditMode)
            If TSCD_new IsNot Nothing Then
                Me.Load_GRV_TSCD()
                'grv_focus()
            End If
        Else
            MessageBox.Show("Không có bản ghi nào", "Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Information)
        End If
    End Sub

    Private Sub Delete()
        If grvTSCD.RowCount <> 0 Then
            Dim i As Integer = grvTSCD.FocusedRowHandle
            'DM_TSCD_ID = Grid_GetDataRowItemString(grvTSCD, "Id")
            DM_TSCD_ID = CType(grvTSCD.GetRow(i), DmTscd).Id

            If (Grid_GetDataRowItemInteger(grvTSCD, "GIAM_TSCD") = 1) Then
                Message_Information("Tài sản đã khai báo giảm")
                Exit Sub
            End If
            If (MessageBox.Show("Bạn có chắc chắn muốn xóa bản ghi này?", "Thông báo", MessageBoxButtons.YesNo, MessageBoxIcon.Question) = Windows.Forms.DialogResult.Yes) Then
                DmTscdService.DeleteById(DM_TSCD_ID)
                Me.Load_GRV_TSCD()
            End If
        Else
            MessageBox.Show("Không có bản ghi nào", "Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Information)
        End If


    End Sub

    Private Sub grv_focus()
        'TNCommon.GridControl_SetRowFocus(grcTSCD, "MaTscd", TSCD_new.MaTscd)
    End Sub
#End Region

#Region "Events"

    Private Sub frmThongTinTaiSan_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Try
            'btnPrint.Visible = False

            BindData()
            Dim F_ngay As Date = Date.Now.Date
            Dim S_ngay As Date = New Date(F_ngay.Year, 1, 1)
            dteTuNgay.DateTime = S_ngay
            dteDenNgay.DateTime = F_ngay
            Load_GRV_TSCD()
            grvTSCD.FixedLineWidth = 0
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnCreate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnAdd.Click
        Try
            Create()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnModify_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnModify.Click
        Try
            Edit()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub



    Private Sub frmThongTinTaiSan_DanhSach_KeyDown(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyDown
        Try
            Select Case e.KeyCode
                Case Keys.Escape
                    Me.Close()
                    Exit Select
                Case Keys.F3
                    Create()
                    Exit Select
                Case Keys.F4
                    Edit()
                    Exit Select
                Case Keys.F8
                    Delete()
                    Exit Select
            End Select
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        Try
            Me.Close()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnIn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnPrint.Click
        Try
            Dim frm As In_The_TSCD = ObjectFactory.GetObject("In_The_TSCD", New Object() {dteTuNgay.DateTime, dteDenNgay.DateTime, txtTenTS.Text, grlLoaiTSCD.EditValue, Generals.DON_VI.Id})
            frm.ShowDialog()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnSearch_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSearch.Click
        Try
            frmProgress.Instance.Thread = AddressOf Load_GRV_TSCD
            frmProgress.Instance.Show_Progress()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub grlLoaiTSCD_EditValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles grlLoaiTSCD.EditValueChanged
        Try
            Dim dr As DmLoaitaisan = CType(ComboHelper.GetSelectData(grlLoaiTSCD), DmLoaitaisan)
            If dr IsNot Nothing Then
                txttenLoaiTS.Text = dr.TenLoaitaisan
            End If
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub
#End Region

    Private Sub btnXem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnXem.Click
        Try
            If grvTSCD.RowCount <> 0 Then
                Dim i As Integer = grvTSCD.FocusedRowHandle
                'DM_TSCD_ID = Grid_GetDataRowItemString(grvTSCD, "Id")
                DM_TSCD_ID = CType(grvTSCD.GetRow(i), DmTscd).Id

                Dim frm As New frmThongTinTSCD_ChiTiet
                TSCD_new = frm.Show_Form(DM_TSCD_ID, DataInputState.ViewMode)
                If TSCD_new IsNot Nothing Then
                    Me.Load_GRV_TSCD()
                    grv_focus()
                End If
            Else
                MessageBox.Show("Không có bản ghi nào", "Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Information)
            End If
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub grvTSCD_FocusedRowChanged(ByVal sender As System.Object, ByVal e As DevExpress.XtraGrid.Views.Base.FocusedRowChangedEventArgs) Handles grvTSCD.FocusedRowChanged
        Try
            If Grid_GetDataRowItemInteger(grvTSCD, "GIAM_TSCD") = 1 Then
                btnXem.Enabled = True
                btnModify.Enabled = False
                btnDelete.Enabled = False
            Else
                btnXem.Enabled = False
                btnModify.Enabled = True
                btnDelete.Enabled = True
            End If
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnDelete_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnDelete.Click
        Try
            Delete()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub
End Class

File: src/Asset/Vns.Erp.Core.Asset.CsAssetGUI/Asset/frmThongTinTaiSan_DanhSach.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Windows.Forms;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraGrid.Views.Grid;

using Vns.Erp.Core.Asset.Domain;
using Vns.Erp.Core.Asset.Service.Interface;

namespace Vns.Erp.Core.Asset.CsAssetGUI
{
    public partial class frmThongTinTaiSan_DanhSach
    {

        #region Property
        private IDmLoaitaisanService _DmLoaitaisanService;
        public IDmLoaitaisanService DmLoaitaisanService
        {
            get { return _DmLoaitaisanService; }
            set { _DmLoaitaisanService = value; }
        }
        private IDmTscdService _DmTscdService;
        public IDmTscdService DmTscdService
        {
            get { return _DmTscdService; }
            set { _DmTscdService = value; }
        }
        #endregion

        #region Variables

        Guid DM_TSCD_ID;
        #endregion
        DmTscd TSCD_new;

        #region Private functions

        private void AddEvents()
        {
            grvTSCD.FocusedRowChanged += grvTSCD_FocusedRowChanged;
            btnDelete.Click += btnDelete_Click;
            this.Load += frmThongTinTaiSan_Load;
            btnAdd.Click += btnCreate_Click;
            btnModify.Click += btnModify_Click;
            this.KeyDown += frmThongTinTaiSan_DanhSach_KeyDown;
            btnClose.Click += btnClose_Click;
            btnPrint.Click += btnIn_Click;
            btnSearch.Click += btnSearch_Click;
            grlLoaiTSCD.EditValueChanged += grlLoaiTSCD_EditValueChanged;
            btnXem.Click += btnXem_Click;
        }

        public frmThongTinTaiSan_DanhSach()
        {
            InitializeComponent();
            AddEvents();
            TNCommon.setIconControl(this);
        }

        private void BindData()
        {
            Guid DON_VI_ID = Generals.DON_VI.Id;
            ArrayList objColumnDefine = null;

            //Dim dt_DM_LoaiTS As DataTable = DM_LoaitaisanBO.Instance.GetAll.Tables(0)
            List<DmLoaitaisan> dt_DM_LoaiTS = new List<DmLoaitaisan>();

            dt_DM_LoaiTS.AddRange(_DmLoaitaisanService.GetAllSortByKyHieu());

            DmLoaitaisan dr = new DmLoaitaisan();
            dr.KyHieu = "";
            dr.Id = Null.NullGuid;
            dr.TenLoaitaisan = "<Không xác định>";
            dt_DM_LoaiTS.Insert(0, dr);

            objColumnDefine = new ArrayList();
            objColumnDefine.Add(new LookupEdit_ColumnInfo("KyHieu", "Ký hiệu"));
            objColumnDefine.Add(new LookupEdit_ColumnInfo("TenLoaitaisan", "Tên loại tài sản"));
            FormGlobals.GridLookupEdit_BindData(grlLoaiTSCD, dt_DM_LoaiTS, "KyHieu", "Id", FormGlobals.COMBO_ITEM_BLANK, "TenLoaitaisan", txttenLoaiTS, objColumnDefine);
        }

        //bind data vao grid
        private void Load_GRV_TSCD()
        {
            IList<DmTscd> lst_TSCD = new List<DmTscd>();
            lst_TSCD = _DmTscdService.GetbyParameter((DateTime)dteTuNgay.DateTime, (DateTime)dteDenNgay.DateTime, txtTenTS.Text, VnsConvert.CGuid(grlLoaiTSCD.EditValue), Generals.DON_VI.Id);
            grcTSCD.DataSource = lst_TSCD;
            //grvTSCD.BestFitColumns()
        }

        private void Create()
        {
            frmThongTinTSCD_ChiTiet frm = (frmThongTinTSCD_ChiTiet)ObjectFactory.GetObject("frmThongTinTSCD_ChiTiet");
            TSCD_new = frm.Show_Form(Null.NullGuid, FormGlobals.DataInputState.AddMode);
            if (TSCD_new != null)
            {
                this.Load_GRV_TSCD();
                grv_focus();
            }
        }

        private void Edit()
        {
            if (grvTSCD.RowCount != 0)
            {
                int i = grvTSCD.FocusedRowHandle;
                //DM_TSCD_ID = Grid_GetDataRowItemString(grvTSCD, "Id")
                DM_TSCD_ID = ((DmTscd)grvTSCD.GetRow(i)).Id;

                if ((FormGlobals.Grid_GetDataRowItemInteger(grvTSCD, "GIAM_TSCD") == 1))
                {
                    FormGlobals.Message_Information("Tài sản đã khai báo giảm");
                    return;
                }

                frmThongTinTSCD_ChiTiet frm = (frmThongTinTSCD_ChiTiet)ObjectFactory.GetObject("frmThongTinTSCD_ChiTiet");
                TSCD_new = frm.Show_Form(DM_TSCD_ID, FormGlobals.DataInputState.EditMode);
                if (TSCD_new != null)
                {
                    this.Load_GRV_TSCD();
                    //grv_focus()
                }
            }
            else
            {
                MessageBox.Show("Không có bản ghi nào", "Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
        }

        private void Delete()
        {
            if (grvTSCD.RowCount != 0)
            {
                int i = grvTSCD.FocusedRowHandle;
                //DM_TSCD_ID = Grid_GetDataRowItemString(grvTSCD, "Id")
                DM_TSCD_ID = ((DmTscd)grvTSCD.GetRow(i)).Id;

                if ((FormGlobals.Grid_GetDataRowItemInteger(grvTSCD, "GIAM_TSCD") == 1))
                {
                    FormGlobals.Message_Information("Tài sản đã khai báo giảm");
                    return;
                }
                if ((MessageBox.Show("Bạn có chắc chắn muốn xóa bản ghi này?", "Thông báo", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == System.Windows.Forms.DialogResult.Yes))
                {
                    DmTscdService.DeleteTscd(DM_TSCD_ID);
                    this.Load_GRV_TSCD();
                }
            }
            else
            {
                MessageBox.Show("Không có bản ghi nào", "Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Information);
            }


        }

        private void grv_focus()
        {
            //TNCommon.GridControl_SetRowFocus(grcTSCD, "MaTscd", TSCD_new.MaTscd)
        }
        #endregion

        #region "Events"

        private void frmThongTinTaiSan_Load(System.Object sender, System.EventArgs e)
        {
            try
            {
                //btnPrint.Visible = False

                BindData();
                System.DateTime F_ngay = System.DateTime.Now.Date;
                System.DateTime S_ngay = new System.DateTime(F_ngay.Year, 1, 1);
                dteTuNgay.DateTime = S_ngay;
                dteDenNgay.DateTime = F_ngay;
                Load_GRV_TSCD();
                grvTSCD.FixedLineWidth = 0;
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnCreate_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                Create();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnModify_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                Edit();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }



        private void frmThongTinTaiSan_DanhSach_KeyDown(System.Object sender, System.Windows.Forms.KeyEventArgs e)
        {
            try
            {
                switch (e.KeyCode)
                {
                    case Keys.Escape:
                        this.Close();
                        break; // TODO: might not be correct. Was : Exit Select

                        break;
                    case Keys.F3:
                        Create();
                        break; // TODO: might not be correct. Was : Exit Select

                        break;
                    case Keys.F4:
                        Edit();
                        break; // TODO: might not be correct. Was : Exit Select

                        break;
                    case Keys.F8:
                        Delete();
                        break; // TODO: might not be correct. Was : Exit Select

                        break;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnClose_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                this.Close();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnIn_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                In_The_TSCD frm = (In_The_TSCD) ObjectFactory.GetObject("In_The_TSCD", new object[] {
				dteTuNgay.DateTime,
				dteDenNgay.DateTime,
				txtTenTS.Text,
				grlLoaiTSCD.EditValue,
				Generals.DON_VI.Id
			});
                frm.ShowDialog();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnSearch_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                frmProgress.Instance().Thread = Load_GRV_TSCD;
                frmProgress.Instance().Show_Progress();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void grlLoaiTSCD_EditValueChanged(System.Object sender, System.EventArgs e)
        {
            try
            {
                DmLoaitaisan dr = (DmLoaitaisan)ComboHelper.GetSelectData(grlLoaiTSCD);
                if (dr != null)
                {
                    txttenLoaiTS.Text = dr.TenLoaitaisan;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }
        #endregion

        private void btnXem_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                if (grvTSCD.RowCount != 0)
                {
                    int i = grvTSCD.FocusedRowHandle;
                    //DM_TSCD_ID = Grid_GetDataRowItemString(grvTSCD, "Id")
                    DM_TSCD_ID = ((DmTscd)grvTSCD.GetRow(i)).Id;

                    frmThongTinTSCD_ChiTiet frm = new frmThongTinTSCD_ChiTiet();
                    TSCD_new = frm.Show_Form(DM_TSCD_ID, FormGlobals.DataInputState.ViewMode);
                    if (TSCD_new != null)
                    {
                        this.Load_GRV_TSCD();
                        grv_focus();
                    }
                }
                else
                {
                    MessageBox.Show("Không có bản ghi nào", "Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void grvTSCD_FocusedRowChanged(System.Object sender, DevExpress.XtraGrid.Views.Base.FocusedRowChangedEventArgs e)
        {
            try
            {
                if (FormGlobals.Grid_GetDataRowItemInteger(grvTSCD, "GIAM_TSCD") == 1)
                {
                    btnXem.Enabled = true;
                    btnModify.Enabled = false;
                    btnDelete.Enabled = false;
                }
                else
                {
                    btnXem.Enabled = false;
                    btnModify.Enabled = true;
                    btnDelete.Enabled = true;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnDelete_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                Delete();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }
    }
}